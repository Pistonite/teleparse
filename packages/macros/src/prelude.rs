use regex::Regex;

pub(crate) use proc_macro::TokenStream;
pub(crate) use proc_macro2::Span;
pub(crate) use quote::{format_ident, quote};
pub(crate) use syn::parse_macro_input;
pub(crate) use syn::punctuated::Punctuated;

/// Type to distinct with proc_macro::TokenStream
pub(crate) type TokenStream2 = proc_macro2::TokenStream;

/// Attribute name, to avoid conflicts with other crates
pub(crate) const CRATE: &str = "teleparse";

pub(crate) fn crate_ident() -> syn::Ident {
    match proc_macro_crate::crate_name(CRATE) {
        Ok(proc_macro_crate::FoundCrate::Name(s)) => syn::Ident::new(&s, Span::call_site()),
        _ => syn::Ident::new(CRATE, Span::call_site()),
    }
}

pub(crate) fn expand_with<F>(input: TokenStream, f: F) -> TokenStream
where
    F: FnOnce(&syn::DeriveInput) -> syn::Result<TokenStream2>,
{
    expand_with_args(input, (), |input, _| f(input))
}

pub(crate) fn expand_with_args<A, F>(input: TokenStream, args: A, f: F) -> TokenStream
where
    F: FnOnce(&syn::DeriveInput, A) -> syn::Result<TokenStream2>,
{
    let derive_input = parse_macro_input!(input as syn::DeriveInput);
    let result = f(&derive_input, args);
    from_result_keep_input(quote! {#derive_input}, result)
}

pub(crate) fn expand_with_mut<F>(input: TokenStream, f: F) -> TokenStream
where
    F: FnOnce(&mut syn::DeriveInput) -> syn::Result<TokenStream2>,
{
    expand_with_args_mut(input, (), |input, _| f(input))
}

pub(crate) fn expand_with_args_mut<A, F>(input: TokenStream, args: A, f: F) -> TokenStream
where
    F: FnOnce(&mut syn::DeriveInput, A) -> syn::Result<TokenStream2>,
{
    let mut derive_input = {
        let input = input.clone();
        parse_macro_input!(input as syn::DeriveInput)
    };
    let result = f(&mut derive_input, args);
    from_result_keep_input(quote! {#derive_input}, result)
}

pub(crate) fn parse_crate_attr_meta(
    attr: &syn::Attribute,
) -> syn::Result<Punctuated<syn::Meta, syn::Token![,]>> {
    attr.parse_args_with(Punctuated::<syn::Meta, syn::Token![,]>::parse_terminated)
}

/// Take teleparse attributes from a list of attributes
pub(crate) fn strip_take_attrs(attrs: &mut Vec<syn::Attribute>) -> Vec<syn::Attribute> {
    let take_attrs = {
        let mut out = Vec::with_capacity(attrs.len());
        std::mem::swap(&mut out, attrs);
        out
    };
    let mut out = Vec::new();
    for attr in take_attrs {
        if !attr.path().is_ident(CRATE) {
            // put it back
            attrs.push(attr);
        } else {
            out.push(attr);
        }
    }
    out
}

pub(crate) fn from_result_keep_input(
    input: TokenStream2,
    result: syn::Result<TokenStream2>,
) -> TokenStream {
    match result {
        Ok(output) => output,
        Err(err) => {
            let output = err.into_compile_error();
            quote! {
                #input
                #output
            }
        }
    }
    .into()
}

/// Macro for creating and returning `syn::Error`
macro_rules! syn_error {
    ($tokens:expr, $msg:expr) => {
        return Err(syn::Error::new_spanned($tokens, $msg))
    };
    ($tokens:expr, $($tt:tt)*) => {
        return Err(syn::Error::new_spanned($tokens, format!($($tt)*)))
    };
}
pub(crate) use syn_error;

/// Parse the `teleparse` attribute on the input and return the meta list, also stripping the attribute
pub(crate) fn parse_strip_root_meta_optional(
    input: &mut syn::DeriveInput,
) -> syn::Result<Option<Punctuated<syn::Meta, syn::Token![,]>>> {
    let root_attrs = strip_take_attrs(&mut input.attrs);
    let root_attr = match ensure_one(root_attrs) {
        EnsureOne::None => return Ok(None),
        EnsureOne::More => syn_error!(
            &input.ident,
            "Multiple root {} attributes found! You might want to merge them.",
            CRATE
        ),
        EnsureOne::One(attr) => attr,
    };
    Ok(Some(parse_crate_attr_meta(&root_attr)?))
}

#[must_use]
pub(crate) fn ensure_one<I: IntoIterator>(input: I) -> EnsureOne<I::Item> {
    let mut iter = input.into_iter();
    match iter.next() {
        None => EnsureOne::None,
        Some(first) => match iter.next() {
            None => EnsureOne::One(first),
            Some(_) => EnsureOne::More,
        },
    }
}

pub(crate) enum EnsureOne<T> {
    None,
    One(T),
    More,
}

pub(crate) fn checked_regex_rule(input: &syn::LitStr) -> syn::Result<Regex> {
    let regex = input.value();
    let regex = match Regex::new(&regex) {
        Err(e) => {
            syn_error!(input, e.to_string());
        }
        Ok(x) => x,
    };
    if regex.find("").is_some() {
        syn_error!(
            input,
            "the rule regular expression must not match the empty string"
        );
    }
    Ok(regex)
}

/// Put a block in an anonymous const block
pub(crate) fn anon_const_block(block: TokenStream2) -> TokenStream2 {
    quote! {
        const _: () = {
            #block
        };
    }
}
