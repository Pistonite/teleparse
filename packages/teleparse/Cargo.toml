[package]
name = "teleparse"
version = "0.1.0"
edition = "2021"
description = "teleparse"
license = "MIT"
exclude = ["**/*.expanded.rs"] 
repository = "https://github.com/Pistonite/teleparse"
authors = ["Pistonight <pistonknight@outlook.com>"]

[dependencies]
teleparse-macros = { path = "../macros", version = "=0.0.5" }
deref-derive = "0.1.0"
num = "0.4.3"
thiserror = "2.0.12"
derivative = "2.2.0"
itertools = "0.14.0"
logos = "0.15.0"
serde = { version = "1.0.219", features = ["derive"], optional = true }

[dev-dependencies]
# backtrace-on-stack-overflow = "0.3.0" # doesn't work on windows
macrotest = "1.1.0"
trybuild = "1.0.105"

[features]
default = ["serde"]
serde = ["dep:serde"]
