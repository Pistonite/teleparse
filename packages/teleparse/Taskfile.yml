version: '3'

tasks:
  dev:
    desc: Run tests in watch mode
    cmds:
      - cargo watch -s "cargo test {{.CLI_ARGS}}"

  dev-doc:
    desc: Watch and serve documentation
    cmds:
      - live-server ../../target/doc -p 8080 &
      - cargo watch -s "cargo doc"

  update-macrotest:
    desc: Refresh macro expand test snapshots
    alias: [umacro]
    env:
      MACROTEST: overwrite
    cmds:
      - cargo test --test macrotest

  update-trybuild:
    alias: [utry]
    desc: Refresh compiler output test snapshots
    env:
      TRYBUILD: overwrite
    cmds:
      - cargo test --test trybuild


