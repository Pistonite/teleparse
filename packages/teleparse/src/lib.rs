#![doc = include_str!("../README.md")]

// So macro works in tests, see
// https://github.com/bkchr/proc-macro-crate/issues/14
extern crate self as teleparse;

// re-export user-facing proc-macros
pub use teleparse_macros::{derive_lexicon, derive_syntax};

/// prelude for all traits and common traits when working with this library
pub mod prelude {
    pub use crate::{derive_lexicon, derive_syntax};

    // traits
    pub use crate::Lexer as _;
    pub use crate::Lexicon as _;
    pub use crate::Produce as _;
    pub use crate::Production as _;
    pub use crate::Root as _;
    pub use crate::ToSpan as _;

    pub use crate::tp;
    // util
    pub use crate::{tp::Node, Pos, Span, Token};

    // macros
    pub use crate::{first_set, follow_set, terminal_set, token_set};
    // pub use crate::{derive_root, assert_ll1};
}

pub mod lex;
#[doc(inline)]
pub use lex::{Lexer, Lexicon, Pos, Span, ToSpan, Token};

pub mod syntax;
#[doc(inline)]
pub use syntax::Production;

pub mod parser;
#[doc(inline)]
pub use parser::{Parser, Produce, Root};

mod tp_impl;

pub mod tp {
    pub use crate::tp_impl::option::Exists;
    pub use crate::tp_impl::option::Optional as Option;
    pub use crate::tp_impl::string::{Parse, Quote};
    pub use crate::tp_impl::Node;
    pub type String<T> = Quote<std::string::String, T>;
    pub use crate::tp_impl::iter::Plus;
    pub type Nev<T> = Plus<std::vec::Vec<T>, T>;
    pub type NevDeque<T> = Plus<std::collections::VecDeque<T>, T>;
    pub use crate::tp_impl::iter::Star;
    pub type Vec<T> = Star<std::vec::Vec<T>, T>;
    pub type VecDeque<T> = Star<std::collections::VecDeque<T>, T>;
    pub use crate::tp_impl::iter::Loop;
    pub use crate::tp_impl::punct::Punct;
    pub use crate::tp_impl::recover::Recover;
    pub use crate::tp_impl::split::Split;
}

#[cfg(test)]
pub(crate) mod test;

/// Error when constructing the grammar (i.e. not actually parsing yet).
#[derive(Debug, Clone, PartialEq, thiserror::Error)]
pub enum GrammarError {
    #[error("Left recursion detected in the grammar! Stack: {0}")]
    LeftRecursion(String),
    #[error("The non-terminal `{0}` has a FIRST/FIRST conflict producing either `{1}` or `{2}`. The conflicting terminals are: {3}")]
    FirstFirstConflict(String, String, String, String),
    #[error("The non-terminal `{0}` has a FIRST/FOLLOW conflict producing `{1}` followed by `{2}`. `{1}` can be empty and both can start with: {3}")]
    FirstFollowSeqConflict(String, String, String, String),
    #[error("The non-terminal `{0}` has conflict in its FIRST and FOLLOW sets. The conflicting terminals are: {1}")]
    FirstFollowConflict(String, String),
}

impl GrammarError {
    pub fn left_recursion(stack: &[&str], current: &str) -> Self {
        let message = format!("{} -> {}", stack.join(" -> "), current);
        Self::LeftRecursion(message)
    }
}

// private module for macros
#[doc(hidden)]
pub mod __priv {
    pub use logos;
    pub use teleparse_macros::*;
}
