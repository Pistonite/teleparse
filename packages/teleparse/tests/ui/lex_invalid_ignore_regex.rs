use teleparse::prelude::*;

#[derive_lexicon]
#[teleparse(ignore(r"\"))]
pub enum MyToken {
    Invalid, 
}

fn main() {}
