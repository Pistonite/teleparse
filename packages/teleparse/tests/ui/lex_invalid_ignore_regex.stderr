error: regex parse error:
           \
           ^
       error: incomplete escape sequence, reached end of pattern prematurely
 --> tests/ui/lex_invalid_ignore_regex.rs:4:20
  |
4 | #[teleparse(ignore(r"\"))]
  |                    ^^^^
